src/assets/wise5/components/component-student.component.ts

```diff
@@ -9,18 +17,463 @@ export abstract class ComponentStudent {
+  isSubmitDirty: boolean = false;
+  isSubmit: boolean = false;
+  isDirty: boolean = false;
+  isDisabled: boolean = false;
+  submitCounter: number = 0;
+  latestAnnotations: any;
+  saveMessage = {
+    text: '',
+    time: ''
+  };
+  requestComponentStateSubscription: Subscription;
+  annotationSavedToServerSubscription: Subscription;
+  nodeSubmitClickedSubscription: Subscription;
+  studentWorkSavedToServerSubscription: Subscription;
 
-  constructor(protected NodeService: NodeService) {}
+  constructor(
+    protected AnnotationService: AnnotationService,
+    protected ComponentService: ComponentService,
+    protected ConfigService: ConfigService,
+    protected NodeService: NodeService,
+    protected StudentDataService: StudentDataService,
+    protected UtilService: UtilService
+  ) {}
 
-  ngOnInit() {
+  ngOnInit(): void {
     this.componentId = this.componentContent.id;
+    this.componentType = this.componentContent.type;
+    this.prompt = this.componentContent.prompt;
+    if (!this.isAuthoringComponentPreviewMode()) {
+      this.latestAnnotations = this.AnnotationService.getLatestComponentAnnotations(
+        this.nodeId,
+        this.componentId,
+        this.workgroupId
+      );
+    }
+    this.registerListeners();
+    this.subscribeToRequestComponentState();
+  }
+
+  ngOnDestroy(): void {
+    this.annotationSavedToServerSubscription.unsubscribe();
+    this.nodeSubmitClickedSubscription.unsubscribe();
+    this.requestComponentStateSubscription.unsubscribe();
+    this.studentWorkSavedToServerSubscription.unsubscribe();
+  }
+
+  isAuthoringComponentPreviewMode(): boolean {
+    return this.mode === 'authoringComponentPreview';
+  }
+
+  registerListeners(): void {
+    this.annotationSavedToServerSubscription = this.AnnotationService.annotationSavedToServer$.subscribe(
+      ({ annotation }) => {
+        if (this.isEventTargetThisComponent(annotation)) {
+          this.latestAnnotations = this.AnnotationService.getLatestComponentAnnotations(
+            this.nodeId,
+            this.componentId,
+            this.workgroupId
+          );
+        }
+      }
+    );
+
+    this.nodeSubmitClickedSubscription = this.NodeService.nodeSubmitClicked$.subscribe(
+      ({ nodeId }) => {
+        if (this.nodeId === nodeId) {
+          this.handleNodeSubmit();
+        }
+      }
+    );
+    this.registerStudentWorkSavedToServerListener();
+  }
+
+  isEventTargetThisComponent(args: any): boolean {
+    return this.isForThisComponent(args);
+  }
+
+  isForThisComponent(object: any) {
+    return this.nodeId === object.nodeId && this.componentId === object.componentId;
+  }
+
+  handleNodeSubmit(): void {
+    this.isSubmit = true;
+  }
+
+  registerStudentWorkSavedToServerListener(): void {
+    this.studentWorkSavedToServerSubscription = this.StudentDataService.studentWorkSavedToServer$.subscribe(
+      (args: any) => {
+        this.handleStudentWorkSavedToServer(args);
+      }
```

Comment:
```suggestion
      ({studentWork: any}) => {
        this.handleStudentWorkSavedToServer(studentWork);
      }
```
Can we just pass in the componentState to the handleStudentWorkSavedToServer function here instead of the args object? It looks like nothing downstream uses other fields in args.

Reply:
Yea it looks like there's always only the studentWork field in that object. We should do this in another issue.